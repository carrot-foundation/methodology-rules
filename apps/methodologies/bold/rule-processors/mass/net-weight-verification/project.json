{
  "name": "methodologies-bold-rule-processors-mass-net-weight-verification",
  "$schema": "../../../../../../node_modules/nx/schemas/project-schema.json",
  "tags": ["methodology:bold", "processor:mass", "type:app"],
  "sourceRoot": "apps/methodologies/bold/rule-processors/mass/net-weight-verification/src",
  "projectType": "application",
  "targets": {
    "build-lambda": {},
    "lint": {},
    "package-lambda": {},
    "upload-lambda": {},
    "test": {},
    "ts": {}
  }
}
