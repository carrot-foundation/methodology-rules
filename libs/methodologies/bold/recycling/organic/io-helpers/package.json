{
  "name": "methodologies/bold/recycling/organic/io-helpers",
  "version": "0.0.0"
}
