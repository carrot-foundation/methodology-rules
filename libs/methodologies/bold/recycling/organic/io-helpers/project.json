{
  "name": "methodologies-bold-recycling-organic-io-helpers",
  "$schema": "../../../../../../node_modules/nx/schemas/project-schema.json",
  "tags": ["type:lib", "app:any", "stack:any", "methodology:bold"],
  "sourceRoot": "libs/methodologies/bold/recycling/organic/io-helpers/src",
  "projectType": "library",
  "targets": {
    "lint": {},
    "test": {},
    "ts": {}
  }
}
