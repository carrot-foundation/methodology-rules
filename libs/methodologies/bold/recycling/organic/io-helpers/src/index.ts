export * from './document.helpers';
export * from './document-query.service';
export * from './testing.helpers';
export * from './document-query.service.types';
