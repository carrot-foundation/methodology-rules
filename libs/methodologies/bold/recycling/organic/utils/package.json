{
  "name": "methodologies/bold/recycling/organic/utils",
  "version": "0.0.0"
}
