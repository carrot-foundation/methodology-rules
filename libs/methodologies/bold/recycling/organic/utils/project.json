{
  "name": "methodologies-bold-recycling-organic-utils",
  "$schema": "../../../../../../node_modules/nx/schemas/project-schema.json",
  "tags": ["type:lib", "app:any", "stack:any", "methodology:bold"],
  "sourceRoot": "libs/methodologies/bold/recycling/organic/utils/src",
  "projectType": "library",
  "targets": {
    "test": {},
    "lint": {},
    "ts": {}
  }
}
