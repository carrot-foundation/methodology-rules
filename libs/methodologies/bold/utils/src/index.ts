export * from './event.helpers';
export * from './rule-processors.constants';
export * from './document.mappers';
export * from './typia.validators';
