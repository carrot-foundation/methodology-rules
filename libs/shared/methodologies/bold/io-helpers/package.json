{
  "name": "@carrot-fndn/shared/methodologies/bold/io-helpers",
  "version": "0.0.0"
}
