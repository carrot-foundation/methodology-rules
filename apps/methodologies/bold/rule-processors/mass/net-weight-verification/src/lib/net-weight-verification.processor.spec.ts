import { loadParentDocument } from '@carrot-fndn/methodologies/bold/io-helpers';
import { stubDocument } from '@carrot-fndn/methodologies/bold/testing';
import {
  type Document,
  DocumentEventAttributeName,
} from '@carrot-fndn/methodologies/bold/types';
import { DOCUMENT_NOT_FOUND_RESULT_COMMENT } from '@carrot-fndn/methodologies/bold/utils';
import {
  type RuleInput,
  type RuleOutput,
  RuleOutputStatus,
} from '@carrot-fndn/shared/rule/types';
import { replaceMetadataAttributeValue } from '@carrot-fndn/shared/testing';
import { faker } from '@faker-js/faker';
import { random } from 'typia';

import { NetWeightVerificationProcessor } from './net-weight-verification.processor';
import {
  APPROVED_WEIGHING_MOVE_EVENT_ATTRIBUTES,
  stubApprovedWeighingMoveEvent,
} from './net-weight-verification.stubs';

jest.mock('@carrot-fndn/methodologies/bold/io-helpers');

const { LOAD_NET_WEIGHT, VEHICLE_GROSS_WEIGHT, VEHICLE_WEIGHT } =
  DocumentEventAttributeName;

describe('NetWeightVerificationProcessor', () => {
  const ruleDataProcessor = new NetWeightVerificationProcessor();
  const documentLoaderService = jest.mocked(loadParentDocument);

  it.each([
    {
      document: stubDocument({
        externalEvents: [
          stubApprovedWeighingMoveEvent(),
          stubApprovedWeighingMoveEvent(),
        ],
      }),
      resultStatus: RuleOutputStatus.APPROVED,
      scenario: 'all the attributes with the correct values',
    },
    {
      document: random<Omit<Document, 'externalEvents'>>(),
      resultStatus: RuleOutputStatus.APPROVED,
      scenario: 'no external events',
    },
    {
      document: stubDocument({
        externalEvents: [
          stubApprovedWeighingMoveEvent(),
          replaceMetadataAttributeValue(
            stubApprovedWeighingMoveEvent(),
            VEHICLE_GROSS_WEIGHT,
            `${faker.number.int()}`,
          ),
        ],
      }),
      resultStatus: RuleOutputStatus.REJECTED,
      scenario: 'one of the events has invalid attributes',
    },
    {
      document: stubDocument({
        externalEvents: [
          replaceMetadataAttributeValue(
            stubApprovedWeighingMoveEvent(),
            VEHICLE_GROSS_WEIGHT,
            '12.00 KG',
          ),
        ],
      }),
      resultStatus: RuleOutputStatus.REJECTED,
      scenario:
        'load-net-weight is not equal to vehicle-gross-weight minus vehicle-weight',
    },
    ...[VEHICLE_WEIGHT, VEHICLE_GROSS_WEIGHT, LOAD_NET_WEIGHT].map(
      (attributeName) => ({
        document: stubDocument({
          externalEvents: [
            replaceMetadataAttributeValue(
              stubApprovedWeighingMoveEvent(),
              attributeName,
              `${faker.number.int()}`,
            ),
          ],
        }),
        resultStatus: RuleOutputStatus.REJECTED,
        scenario: `the ${attributeName} attribute is missing the measurement unit`,
      }),
    ),
    ...[VEHICLE_WEIGHT, VEHICLE_GROSS_WEIGHT, LOAD_NET_WEIGHT].map(
      (attributeName) => ({
        document: stubDocument({
          externalEvents: [
            replaceMetadataAttributeValue(
              stubApprovedWeighingMoveEvent(),
              attributeName,
              'undefined KG',
            ),
          ],
        }),
        resultStatus: RuleOutputStatus.REJECTED,
        scenario: `the ${attributeName} attribute is not in the correct format`,
      }),
    ),
    ...[VEHICLE_WEIGHT, VEHICLE_GROSS_WEIGHT, LOAD_NET_WEIGHT].map(
      (attributeName) => ({
        document: stubDocument({
          externalEvents: [
            {
              ...stubApprovedWeighingMoveEvent(),
              metadata: {
                attributes: APPROVED_WEIGHING_MOVE_EVENT_ATTRIBUTES.filter(
                  ({ name }) => name !== attributeName,
                ),
              },
            },
          ],
        }),
        resultStatus: RuleOutputStatus.APPROVED,
        scenario: `the ${attributeName} attribute is not present`,
      }),
    ),
  ])(
    `should return $resultStatus when the document has $scenario`,
    async ({ document, resultStatus }) => {
      const ruleInput = random<Required<RuleInput>>();

      documentLoaderService.mockResolvedValueOnce(document);

      const ruleOutput = await ruleDataProcessor.process(ruleInput);

      const expectedRuleOutput: RuleOutput = {
        requestId: ruleInput.requestId,
        responseToken: ruleInput.responseToken,
        responseUrl: ruleInput.responseUrl,
        resultComment: expect.any(String),
        resultContent: undefined,
        resultStatus,
      };

      expect(ruleOutput.resultComment).toMatchSnapshot();
      expect(ruleOutput).toEqual(expectedRuleOutput);
    },
  );

  it('should return REJECTED if there is no document', async () => {
    const ruleInput = random<Required<RuleInput>>();

    documentLoaderService.mockResolvedValueOnce(undefined);

    const ruleOutput = await ruleDataProcessor.process(ruleInput);

    const expectedRuleOutput: RuleOutput = {
      requestId: ruleInput.requestId,
      responseToken: ruleInput.responseToken,
      responseUrl: ruleInput.responseUrl,
      resultComment: DOCUMENT_NOT_FOUND_RESULT_COMMENT,
      resultStatus: RuleOutputStatus.REJECTED,
    };

    expect(ruleOutput).toEqual(expectedRuleOutput);
  });
});
