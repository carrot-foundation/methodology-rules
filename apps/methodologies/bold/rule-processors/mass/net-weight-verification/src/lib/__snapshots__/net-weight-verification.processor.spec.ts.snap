// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetWeightVerificationProcessor should return APPROVED when the document has all the attributes with the correct values 1`] = `"Weighing MOVE event (1): 128901 - 19700 = 109201 | Weighing MOVE event (2): 128901 - 19700 = 109201"`;

exports[`NetWeightVerificationProcessor should return APPROVED when the document has no external events 1`] = `"Weighing MOVE events not found"`;

exports[`NetWeightVerificationProcessor should return APPROVED when the document has the load-net-weight attribute is not present 1`] = `"Weighing MOVE event (1) is missing attributes: load-net-weight"`;

exports[`NetWeightVerificationProcessor should return APPROVED when the document has the vehicle-gross-weight attribute is not present 1`] = `"Weighing MOVE event (1) is missing attributes: vehicle-gross-weight"`;

exports[`NetWeightVerificationProcessor should return APPROVED when the document has the vehicle-weight attribute is not present 1`] = `"Weighing MOVE event (1) is missing attributes: vehicle-weight"`;

exports[`NetWeightVerificationProcessor should return REJECTED when the document has load-net-weight is not equal to vehicle-gross-weight minus vehicle-weight 1`] = `"Weighing MOVE event (1): Invalid load net weight (12 - 19700 = -19688) should be equal to 109201"`;

exports[`NetWeightVerificationProcessor should return REJECTED when the document has one of the events has invalid attributes 1`] = `"Weighing MOVE event (2) has invalid attributes: Missing measurement unit "KG" on attribute "vehicle-gross-weight"."`;

exports[`NetWeightVerificationProcessor should return REJECTED when the document has the load-net-weight attribute is missing the measurement unit 1`] = `"Weighing MOVE event (1) has invalid attributes: Missing measurement unit "KG" on attribute "load-net-weight"."`;

exports[`NetWeightVerificationProcessor should return REJECTED when the document has the load-net-weight attribute is not in the correct format 1`] = `"Weighing MOVE event (1) has invalid attributes: Invalid value "undefined KG" on attribute "load-net-weight"."`;

exports[`NetWeightVerificationProcessor should return REJECTED when the document has the vehicle-gross-weight attribute is missing the measurement unit 1`] = `"Weighing MOVE event (1) has invalid attributes: Missing measurement unit "KG" on attribute "vehicle-gross-weight"."`;

exports[`NetWeightVerificationProcessor should return REJECTED when the document has the vehicle-gross-weight attribute is not in the correct format 1`] = `"Weighing MOVE event (1) has invalid attributes: Invalid value "undefined KG" on attribute "vehicle-gross-weight"."`;

exports[`NetWeightVerificationProcessor should return REJECTED when the document has the vehicle-weight attribute is missing the measurement unit 1`] = `"Weighing MOVE event (1) has invalid attributes: Missing measurement unit "KG" on attribute "vehicle-weight"."`;

exports[`NetWeightVerificationProcessor should return REJECTED when the document has the vehicle-weight attribute is not in the correct format 1`] = `"Weighing MOVE event (1) has invalid attributes: Invalid value "undefined KG" on attribute "vehicle-weight"."`;
